tests/integration/testvalidator.nim

```diff
@@ -0,0 +1,313 @@
+        # .withLogTopics("sales", "onchain")
+        .some,
+
+    validators:
+      CodexConfigs.init(nodes=2)
+        .withValidationGroups(groups = 2)
+        .withValidationGroupIndex(idx = 0, groupIndex = 0)
+        .withValidationGroupIndex(idx = 1, groupIndex = 1)
+        # .debug() # uncomment to enable console log output
+        .withLogFile() # uncomment to output log file to tests/integration/logs/<start_datetime> <suite_name>/<test_name>/<node_role>_<node_idx>.log
+        .withLogTopics("validator") # each topic as a separate string argument
+        .some
+  ):
+    let client0 = clients()[0].client
+    let expiry = 5.periods
+    let duration = expiry + 10.periods
+
+    # let mine a block to sync the blocktime with the current clock
+    discard await ethProvider.send("evm_mine")
+
+    var currentTime = await ethProvider.currentTime()
+    let requestEndTime = currentTime.truncate(uint64) + duration
+
+    let data = await RandomChunker.example(blocks=8)
+    
+    # TODO: better value for data.len below. This TODO is also present in
+    # testproofs.nim - we may want to address it or remove the comment.
+    createAvailabilities(data.len * 2, duration)
+
+    await marketplace.startTrackingEvents()
+
+    let cid = client0.upload(data).get
+    let purchaseId = await client0.requestStorage(
+      cid,
+      expiry=expiry,
+      duration=duration,
+      nodes=nodes,
+      tolerance=tolerance,
+      proofProbability=proofProbability
+    )
+    let requestId = client0.requestId(purchaseId).get
+
+    debug "validation suite", purchaseId = purchaseId.toHex, requestId = requestId
+
+    check eventuallyS(client0.purchaseStateIs(purchaseId, "started"),
+      timeout = (expiry + 60).int, step = 5)
+    
+    # if purchase state is not "started", it does not make sense to continue
+    without purchaseState =? client0.getPurchase(purchaseId).?state:
+      fail()
+      return
+    
+    debug "validation suite", purchaseState = purchaseState 
+
+    if purchaseState != "started":
+      fail()
+      return
+
+    discard await ethProvider.send("evm_mine")
+    currentTime = await ethProvider.currentTime()
+    let secondsTillRequestEnd = (requestEndTime - currentTime.truncate(uint64)).int
+
+    debug "validation suite", secondsTillRequestEnd = secondsTillRequestEnd.seconds
+    
+    # Because of erasure coding, after (tolerance + 1) slots are freed, the
+    # remaining nodes are be freed but marked as "Failed" as the whole
+    # request fails. A couple of checks to capture this:
+    let expectedSlotsFreed = tolerance + 1
+    check eventuallyS(checkSlotsFreed(requestId, expectedSlotsFreed),
+      timeout = secondsTillRequestEnd + 60, step = 5,
+      cancelExpression = isRequestCancelled(requestId))
+    
+    # extra check
+    await marketplace.checkSlotsFailed(requestId)
```

Comment:
By checking that the remaining slots, for which we did not capture `slotFreed` event, have status failed, we de-facto confirm that the request has failed.
This extra check documents the logic (that in the end the request should fail), and does not cost anything, so why not?

Reply:
The reason the extra slots would have a `slotState` of `failed` is because the request state is `Failed`. Having contract event emissions not aligned with request and slot states is not something we should be testing here, and should leave it to testing inside codex-contracts-eth and formal verification.  We do not need to check contract events to be sure of this.

Also, keeping harness logic to a minimum means the less of a chance we have of introducing a bug in the harness itself.

Reply:
Keeping things simple sounds always convincing, that's why I am willing to check the `marketplace.getRequest` path. I see your point that perhaps saying myself that I want to get the sensing as much narrow as possible and closest to the source as possible, I start seeing that perhaps this extra check is not fully justifiable here...

Let me try to refine that a bit.

Reply:
Done as described in https://github.com/codex-storage/nim-codex/pull/922#discussion_r1870519963